text/0000-public-private-dependencies.md

```diff
@@ -0,0 +1,192 @@
+- Feature Name: `public_private_dependencies`
+- Start Date: 2017-04-03
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Introduce a public/private distinction to crate dependencies.
+
+# Motivation
+[motivation]: #motivation
+
+The crates ecosystem has greatly expanded since Rust 1.0 and with that a few patterns for
+dependencies have evolved that challenge the currently existing dependency declaration
+system in cargo and rust.  The most common problem is that a crate `A` depends on
+another crate `B` but some of the types from crate `B` are exposed through the API in
+crate `A`.  This causes problems in practice if that dependency `B` is also used by the
+user's code itself which often leaves users in less than ideal situations where either
+their code refuses to compile because different versions of those libraries are requested
+or where compiler messages are less than clear.
+
+The introduction of an explicit distinction between public and private dependencies can
+solve some of these issues and also let us lift some restrictions that should make some
+code compile that previously was prevented from compiling by restrictions in cargo.
+
+**Q: What is a public dependency?**<br>
+A: a dependency is public if some of the types or trait of that dependency is itself
+exported through the main crate.  The most common places where this happens is obviously
+return values and function parameter but obviously the same applies to trait implementations
+and many other things.  Because public can be tricky to determine for a user this RFC
+proposes to extend the compiler infrastructure to detect the concept of "public dependency".
+This will help the user understanding this concept and avoid making mistakes in
+the `Cargo.toml`
+
+Effectively the idea is that if your own library bumps a public dependency it means that
+it's a breaking change of your *own* crate.
+
+**Q: What is a private dependency?**<br>
+A: On the other hand a private dependency is contained within your crate and effectively
+invisible for users of your crate.  As a result private dependencies can be freely
+duplicated.  This distinction will also make it possible to relax some restrictions that
+currently exist in Cargo which sometimes prevent crates from compiling.
+
+**Q: Can public become private later?**<br>
+A: Public dependencies are public within a reachable subgraph but can become private if a
+crate stops exposing a public dependency.  For instance it is very possible to have a
+family of crates that all depend on a utility crate that provides common types which is
+a public dependency for all of them.  However your own crate only becomes a user of this
```

Comment:
This paragraph has confusing wording, at least to me. To reiterate the point of this example, you could say that "public" is a property of the dependency relation (the edge in the dependency graph) and not of the crate or the package (the node).